src/coprocessor/dag/batch/executors/top_n_executor.rs

```diff
@@ -0,0 +1,725 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cmp::Ordering;
+use std::collections::BinaryHeap;
+use std::pin::Pin;
+use std::ptr::NonNull;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use tipb::executor::TopN;
+use tipb::expression::{Expr, FieldType};
+
+use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+use crate::coprocessor::codec::data_type::VectorValue;
+use crate::coprocessor::dag::batch::interface::*;
+use crate::coprocessor::dag::expr::EvalWarnings;
+use crate::coprocessor::dag::expr::{EvalConfig, EvalContext};
+use crate::coprocessor::dag::rpn_expr::types::RpnStackNode;
+use crate::coprocessor::dag::rpn_expr::{RpnExpression, RpnExpressionBuilder};
+use crate::coprocessor::Result;
+
+pub struct BatchTopNExecutor<Src: BatchExecutor> {
+    /// The heap, which contains N rows at most.
+    ///
+    /// This field is placed before `eval_columns_buffer_unsafe`, `order_exprs`, `order_is_desc`
+    /// and `src` because it relies on data in those fields and we want this field to be dropped
+    /// first.
+    heap: BinaryHeap<HeapItemUnsafe>,
+
+    /// A collection of all evaluated columns. This is to avoid repeated allocations in
+    /// each `next_batch()`.
+    ///
+    /// DO NOT EVER try to read the content of the elements directly, since it is highly unsafe.
+    /// The lifetime of elements is not really 'static. Certain elements are valid only if both
+    /// of the following conditions are satisfied:
+    ///
+    /// 1. `BatchTopNExecutor` is valid (i.e. not dropped).
+    ///
+    /// 2. The referenced `LazyBatchColumnVec` of the element must be valid, which only happens
+    ///    when at least one of the row is in the `heap`. Note that rows may be swapped out from
+    ///    `heap` at any time.
+    ///
+    /// This field is placed before `order_exprs` and `src` because it relies on data in
+    /// those fields and we want this field to be dropped first.
+    #[allow(clippy::box_vec)]
+    eval_columns_buffer_unsafe: Pin<Box<Vec<RpnStackNode<'static>>>>,
+
+    order_exprs: Pin<Vec<RpnExpression>>,
+
+    /// Whether or not it is descending order for each order by column.
+    order_is_desc: Pin<Vec<bool>>,
+
+    n: usize,
+
+    context: EvalContext,
+    src: Src,
+    is_ended: bool,
+}
+
+unsafe impl<Src: BatchExecutor + Send> Send for BatchTopNExecutor<Src> {}
```

Comment:
For reviewers: This character needs to be verified carefully. Please help confirm `Send` bound can be satisfied. @sticnarf @lonng